File: №4 Abstract Factory Pattern/Java/src/Main.java

abstract class Ticket {
    public abstract void display();
}
class AdultTicket extends Ticket {
    @Override
    public void display() {
        System.out.println("Adult Ticket");
    }
}
class ChildTicket extends Ticket {
    @Override
    public void display() {
        System.out.println("Child Ticket");
    }
}
class SeniorTicket extends Ticket {
    @Override
    public void display() {
        System.out.println("Senior Ticket");
    }
}

abstract class PaymentMethod {
    public abstract void processPayment();
}

class CreditCardPayment extends PaymentMethod {
    @Override
    public void processPayment() {
        System.out.println("Processing payment with Credit Card");
    }
}

class CashPayment extends PaymentMethod {
    @Override
    public void processPayment() {
        System.out.println("Processing payment with Cash");
    }
}

interface TicketFactory {
    Ticket createTicket();
    PaymentMethod createPaymentMethod();
}

class AdultTicketFactory implements TicketFactory {
    @Override
    public Ticket createTicket() {
        return new AdultTicket();
    }

    @Override
    public PaymentMethod createPaymentMethod() {
        return new CreditCardPayment();
    }
}

class ChildTicketFactory implements TicketFactory {
    @Override
    public Ticket createTicket() {
        return new ChildTicket();
    }

    @Override
    public PaymentMethod createPaymentMethod() {
        return new CashPayment();
    }
}

class SeniorTicketFactory implements TicketFactory {
    @Override
    public Ticket createTicket() {
        return new SeniorTicket();
    }

    @Override
    public PaymentMethod createPaymentMethod() {
        return new CreditCardPayment();
    }
}

public class Main {
    public static void main(String[] args) {
        TicketFactory adultTicketFactory = new AdultTicketFactory();
        TicketFactory childTicketFactory = new ChildTicketFactory();
        TicketFactory seniorTicketFactory = new SeniorTicketFactory();

        Ticket adultTicket = adultTicketFactory.createTicket();
        PaymentMethod adultPaymentMethod = adultTicketFactory.createPaymentMethod();

        Ticket childTicket = childTicketFactory.createTicket();
        PaymentMethod childPaymentMethod = childTicketFactory.createPaymentMethod();

        Ticket seniorTicket = seniorTicketFactory.createTicket();
        PaymentMethod seniorPaymentMethod = seniorTicketFactory.createPaymentMethod();

        adultTicket.display();
        adultPaymentMethod.processPayment();

        childTicket.display();
        childPaymentMethod.processPayment();

        seniorTicket.display();
        seniorPaymentMethod.processPayment();
    }
}

File: №4 Abstract Factory Pattern/Python/main.py
class Ticket:
    def display(self):
        pass

class AdultTicket(Ticket):
    def display(self):
        print("Adult Ticket")

class ChildTicket(Ticket):
    def display(self):
        print("Child Ticket")

class SeniorTicket(Ticket):
    def display(self):
        print("Senior Ticket")

class PaymentMethod:
    def process_payment(self):
        pass

class CreditCardPayment(PaymentMethod):
    def process_payment(self):
        print("Processing payment with Credit Card")

class CashPayment(PaymentMethod):
    def process_payment(self):
        print("Processing payment with Cash")

class TicketFactory:
    def create_ticket(self):
        pass

    def create_payment_method(self):
        pass

class AdultTicketFactory(TicketFactory):
    def create_ticket(self):
        return AdultTicket()

    def create_payment_method(self):
        return CreditCardPayment()

class ChildTicketFactory(TicketFactory):
    def create_ticket(self):
        return ChildTicket()

    def create_payment_method(self):
        return CashPayment()

class SeniorTicketFactory(TicketFactory):
    def create_ticket(self):
        return SeniorTicket()

    def create_payment_method(self):
        return CreditCardPayment()

if __name__ == "__main__":
    adult_ticket_factory = AdultTicketFactory()
    child_ticket_factory = ChildTicketFactory()
    senior_ticket_factory = SeniorTicketFactory()

    adult_ticket = adult_ticket_factory.create_ticket()
    adult_payment_method = adult_ticket_factory.create_payment_method()

    child_ticket = child_ticket_factory.create_ticket()
    child_payment_method = child_ticket_factory.create_payment_method()

    senior_ticket = senior_ticket_factory.create_ticket()
    senior_payment_method = senior_ticket_factory.create_payment_method()

    adult_ticket.display()
    adult_payment_method.process_payment()

    child_ticket.display()
    child_payment_method.process_payment()

    senior_ticket.display()
    senior_payment_method.process_payment()
